In [2]:
import pandas as pd
import numpy as np
demog_df=pd.read_csv("C:\\Users\\al_ce\\Documents\\CELIA\\master data analytics\\PROYECTO_FINAL\\DATA\\nutrition_raw\\demographic.csv")
diet_df=pd.read_csv("C:\\Users\\al_ce\\Documents\\CELIA\\master data analytics\\PROYECTO_FINAL\\DATA\\nutrition_raw\\diet.csv")
exam_df=pd.read_csv("C:\\Users\\al_ce\\Documents\\CELIA\\master data analytics\\PROYECTO_FINAL\\DATA\\nutrition_raw\\examination.csv")
question_df=pd.read_csv("C:\\Users\\al_ce\\Documents\\CELIA\\master data analytics\\PROYECTO_FINAL\\DATA\\nutrition_raw\\questionnaire.csv")


In [3]:
#se comprueba el número de filas y columnas de cada dataset
print(demog_df.shape,
diet_df.shape,
exam_df.shape,
question_df.shape)

(10175, 47) (9813, 168) (9813, 224) (10175, 953)


## 1. Limpieza del dataset demographic

In [4]:
# Ánalisis general: visualización de las primeras filas para la comprensión de datos
demog_df.head()

,SEQN,SDDSRVYR,RIDSTATR,RIAGENDR,RIDAGEYR,RIDAGEMN,RIDRETH1,RIDRETH3,RIDEXMON,RIDEXAGM,...,DMDHREDU,DMDHRMAR,DMDHSEDU,WTINT2YR,WTMEC2YR,SDMVPSU,SDMVSTRA,INDHHIN2,INDFMIN2,INDFMPIR
0,73557,8,2,1,69,NaN,4,4,1.0,NaN,...,3.0,4.0,NaN,13281.237386,13481.042095,1,112,4.0,4.0,0.84
1,73558,8,2,1,54,NaN,3,3,1.0,NaN,...,3.0,1.0,1.0,23682.057386,24471.769625,1,108,7.0,7.0,1.78
2,73559,8,2,1,72,NaN,3,3,2.0,NaN,...,4.0,1.0,3.0,57214.803319,57193.285376,1,109,10.0,10.0,4.51
3,73560,8,2,1,9,NaN,3,3,1.0,119.0,...,3.0,1.0,4.0,55201.178592,55766.512438,2,109,9.0,9.0,2.52
4,73561,8,2,2,73,NaN,3,3,1.0,NaN,...,5.0,1.0,5.0,63709.667069,65541.871229,2,116,15.0,15.0,5.00


### 1.1 Transformación de columnas demográficas

In [5]:
demog_df.columns

Index(['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN',
       'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'RIDEXAGM', 'DMQMILIZ', 'DMQADFC',
       'DMDBORN4', 'DMDCITZN', 'DMDYRSUS', 'DMDEDUC3', 'DMDEDUC2', 'DMDMARTL',
       'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG', 'FIAPROXY',
       'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA', 'DMDHHSIZ',
       'DMDFMSIZ', 'DMDHHSZA', 'DMDHHSZB', 'DMDHHSZE', 'DMDHRGND', 'DMDHRAGE',
       'DMDHRBR4', 'DMDHREDU', 'DMDHRMAR', 'DMDHSEDU', 'WTINT2YR', 'WTMEC2YR',
       'SDMVPSU', 'SDMVSTRA', 'INDHHIN2', 'INDFMIN2', 'INDFMPIR'],
      dtype='object')

De las 47 variables demográficas originales se seleccionan aquellas relevantes para el ánalisis, reduciendo dimensionalidad. Se renombran las columnas seleccionadas para mejorar la interpretabilidad.

In [6]:
#se renombran las columnas
demog_df=demog_df.rename(columns={
    "SEQN": "person_id",
    "RIAGENDR": "gender",
    "RIDAGEYR": "age",
    "RIDRETH1": "race",
    "DMDBORN4": "country",
    "DMDEDUC2": "education_level",
    "DMDMARTL": "marital_status",
    "RIDEXPRG": "pregnancy_status",
    "DMDHHSIZ": "household_size",
    "INDHHIN2": "household_income",
    "INDFMIN2": "family_income",
    "INDFMPIR": "family_poverty_ratio"})

In [7]:
#selección de las columnas relevantes
demog_df=demog_df[[
    "person_id",
    "gender",
    "age",
    "race",
    "country",
    "education_level",
    "marital_status",
    "pregnancy_status",
    "household_size",
    "household_income",
    "family_income",
    "family_poverty_ratio"]]

In [8]:
#comprobamos
demog_df.head()

,person_id,gender,age,race,country,education_level,marital_status,pregnancy_status,household_size,household_income,family_income,family_poverty_ratio
0,73557,1,69,4,1,3.0,4.0,NaN,3,4.0,4.0,0.84
1,73558,1,54,3,1,3.0,1.0,NaN,4,7.0,7.0,1.78
2,73559,1,72,3,1,4.0,1.0,NaN,2,10.0,10.0,4.51
3,73560,1,9,3,1,NaN,NaN,NaN,4,9.0,9.0,2.52
4,73561,2,73,3,1,5.0,1.0,NaN,2,15.0,15.0,5.00


Se observa que en las columnas 'gender', 'race', 'country', 'eduaction_level', 'marital_status' y 'pregnancy_status' aparecen valores numéricos que dificultan la comprensión de los datos. 
Además, las columnas 'household_income', 'family_income' y 'poverty_ratio' contienen valores numéricos codificados.
Por ello, se aplican mapeos de código a texto.

In [9]:
# Mapeamos códigos a texto
demog_df['gender']=demog_df['gender'].map({1: 'Male', 2: 'Female'})
demog_df['race']=demog_df['race'].map({
    1: 'Mexican American',
    2: 'Other Hispanic',
    3: 'Non-Hispanic White',
    4: 'Non-Hispanic Black',
    5: 'Other/Multi-Racial'
})
demog_df['country']=demog_df['country'].map({
    1: 'USA',
    2: 'Mexico',
    3: 'Other Spanish',
    4: 'Other non-Spanish'
})
demog_df['education_level']=demog_df['education_level'].map({
    1: 'Less than 9th grade',
    2: '9-11th grade',
    3: 'High school/GED',
    4: 'Some college/AA',
    5: 'College graduate or above'
})
demog_df['marital_status']=demog_df['marital_status'].map({
    1: 'Married',
    2: 'Widowed',
    3: 'Divorced',
    4: 'Separated',
    5: 'Never married',
    6: 'Living with partner'
})
demog_df['pregnancy_status']=demog_df['pregnancy_status'].map({1: 'Yes', 2: 'No'})
demog_df['household_income']=demog_df['household_income'].map({
    1:	' 0 to 4,999 $'	,
    2:	'5,000 to 9,999	$',
    3:	'10,000 to 14,999 $',		
    4:	'15,000 to 19,999 $',		
    5:	'20,000 to 24,999 $',		
    6:	'25,000 to 34,999 $',		
    7:	'35,000 to 44,999 $',	
	8:	'45,000 to 54,999 $',	
    9:	'55,000 to 64,999 $',		
    10:	'65,000 to 74,999 $',		
    12:	'Over 20,000 $',	
    13:	'Under 20,000 $',	
    14:	'75,000 to 99,999 $',		
	15:	'100,000 $ and Over',	
    77:	'Refused',		
	99:	'Dont know'
})



In [10]:
#Comprobamos
demog_df.head()

,person_id,gender,age,race,country,education_level,marital_status,pregnancy_status,household_size,household_income,family_income,family_poverty_ratio
0,73557,Male,69,Non-Hispanic Black,USA,High school/GED,Separated,NaN,3,"15,000 to 19,999 $",4.0,0.84
1,73558,Male,54,Non-Hispanic White,USA,High school/GED,Married,NaN,4,"35,000 to 44,999 $",7.0,1.78
2,73559,Male,72,Non-Hispanic White,USA,Some college/AA,Married,NaN,2,"65,000 to 74,999 $",10.0,4.51
3,73560,Male,9,Non-Hispanic White,USA,NaN,NaN,NaN,4,"55,000 to 64,999 $",9.0,2.52
4,73561,Female,73,Non-Hispanic White,USA,College graduate or above,Married,NaN,2,"100,000 $ and Over",15.0,5.00


In [11]:
demog_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10175 entries, 0 to 10174
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   person_id             10175 non-null  int64  
 1   gender                10175 non-null  object 
 2   age                   10175 non-null  int64  
 3   race                  10175 non-null  object 
 4   country               10170 non-null  object 
 5   education_level       5762 non-null   object 
 6   marital_status        5766 non-null   object 
 7   pregnancy_status      1215 non-null   object 
 8   household_size        10175 non-null  int64  
 9   household_income      10042 non-null  object 
 10  family_income         10052 non-null  float64
 11  family_poverty_ratio  9390 non-null   float64
dtypes: float64(2), int64(3), object(7)
memory usage: 954.0+ KB


### 1.2 Tratamiento de valores nulos

Atendiendo al output del código anterior, se observa que no todas las columnas contienen valores nulos. Para las columnas categóricas, se realiza una imputación con 'unknown'; para las columnas numéricas, al no haber un número muy elevado de nulos, se imputa con la mediana. 

In [12]:
#imputamos con unknown en las columnas categóricas
categoric_column=['country', 'education_level', 'marital_status', 'pregnancy_status','household_income']
demog_df[categoric_column] = demog_df[categoric_column].fillna('Unknown')

#imputamos con la mediana en las columnas numéricas
#para no ir una por una creamos una lista
numeric_column=['household_size','family_income','family_poverty_ratio']
demog_df[numeric_column] = demog_df[numeric_column].fillna(demog_df[numeric_column].median())

In [13]:
#comprobamos que no hay valores nulos
demog_df.isnull().sum()

person_id               0
gender                  0
age                     0
race                    0
country                 0
education_level         0
marital_status          0
pregnancy_status        0
household_size          0
household_income        0
family_income           0
family_poverty_ratio    0
dtype: int64

### 1.3 Comprobación de los valores únicos

In [14]:
demog_df['person_id'].nunique()

10175

## 2. Limpieza del dataset diet

In [15]:
# Ánalisis general: visualización de las primeras filas para la comprensión de datos
diet_df.head()

,SEQN,WTDRD1,WTDR2D,DR1DRSTZ,DR1EXMER,DRABF,DRDINT,DR1DBIH,DR1DAY,DR1LANG,...,DRD370QQ,DRD370R,DRD370RQ,DRD370S,DRD370SQ,DRD370T,DRD370TQ,DRD370U,DRD370UQ,DRD370V
0,73557,16888.327864,12930.890649,1,49.0,2.0,2.0,6.0,2.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,73558,17932.143865,12684.148869,1,59.0,2.0,2.0,4.0,1.0,1.0,...,NaN,2.0,NaN,2.0,NaN,2.0,NaN,2.0,NaN,2.0
2,73559,59641.812930,39394.236709,1,49.0,2.0,2.0,18.0,6.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,73560,142203.069917,125966.366442,1,54.0,2.0,2.0,21.0,3.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,73561,59052.357033,39004.892993,1,63.0,2.0,2.0,18.0,1.0,1.0,...,NaN,2.0,NaN,2.0,NaN,2.0,NaN,2.0,NaN,2.0


### 2.1 Transformación de columnas dietéticas

In [16]:
#obtenemos el nombre de todas las columnas
diet_df.columns

Index(['SEQN', 'WTDRD1', 'WTDR2D', 'DR1DRSTZ', 'DR1EXMER', 'DRABF', 'DRDINT',
       'DR1DBIH', 'DR1DAY', 'DR1LANG',
       ...
       'DRD370QQ', 'DRD370R', 'DRD370RQ', 'DRD370S', 'DRD370SQ', 'DRD370T',
       'DRD370TQ', 'DRD370U', 'DRD370UQ', 'DRD370V'],
      dtype='object', length=168)

Se selccionan las columnas relevantes para el análisis y se renombran para facilitar el manejo y comprensión de los datos.

In [17]:
#se renombran las columnas
diet_df=diet_df.rename(columns={
    "SEQN": "person_id",
    "DR1TKCAL": "kcal",
    "DR1TPROT": "protein_g",
    "DR1TTFAT": "fat_g",
    "DR1TCHOL": "cholesterol_mg",
    "DR1TCARB": "carb_g",
    "DR1TSUGR": "sugar_g",
    "DR1TFIBE": "fiber_g",
    "DR1TALCO": "alcohol_g",
    "DR1TMFAT": "total_monosat_fatty_acids_g",
    "DR1TPFAT": "total_polysat_fatty_acids_g",
    "DR1TCALC": "calcium_mg",
    "DR1TMAGN": "magnesium_mg" ,
    
     
      })

In [18]:
#selección de las columnas relevantes
diet_df=diet_df[[
    "person_id",
    "kcal",
    "protein_g",
    "fat_g",
    "cholesterol_mg",
    "carb_g",
    "sugar_g",
    "fiber_g",
    "alcohol_g",
    "total_monosat_fatty_acids_g",
    "total_polysat_fatty_acids_g",
    "calcium_mg",
    "magnesium_mg"
    ]]

In [19]:
#comprobamos
diet_df.head()

,person_id,kcal,protein_g,fat_g,cholesterol_mg,carb_g,sugar_g,fiber_g,alcohol_g,total_monosat_fatty_acids_g,total_polysat_fatty_acids_g,calcium_mg,magnesium_mg
0,73557,1574.0,43.63,52.81,209.0,239.59,176.47,10.8,0.0,18.493,8.829,949.0,206.0
1,73558,5062.0,338.13,124.29,2584.0,423.78,44.99,16.7,119.0,35.481,20.505,3193.0,903.0
2,73559,1743.0,64.61,65.97,88.0,224.39,102.90,9.9,0.0,20.902,12.953,877.0,210.0
3,73560,1490.0,77.75,58.27,187.0,162.92,80.58,10.6,0.0,16.986,9.617,1521.0,183.0
4,73561,1421.0,55.24,55.36,41.0,178.20,87.78,12.3,0.0,26.216,1.263,1410.0,396.0


In [20]:
#obtenemos información general de la tabla
diet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   person_id                    9813 non-null   int64  
 1   kcal                         8531 non-null   float64
 2   protein_g                    8531 non-null   float64
 3   fat_g                        8531 non-null   float64
 4   cholesterol_mg               8531 non-null   float64
 5   carb_g                       8531 non-null   float64
 6   sugar_g                      8531 non-null   float64
 7   fiber_g                      8531 non-null   float64
 8   alcohol_g                    8531 non-null   float64
 9   total_monosat_fatty_acids_g  8531 non-null   float64
 10  total_polysat_fatty_acids_g  8531 non-null   float64
 11  calcium_mg                   8531 non-null   float64
 12  magnesium_mg                 8531 non-null   float64
dtypes: float64(12), in

### 2.2 Tratamiento de valores nulos

Al obtener la información general de los datos dietéticos, se observa que en todas las columnas (a excepción del identificador de los sujetos) se tiene el mismo número de nulos. A continuación, se analiza si los nulos coinciden en la misma fila, lo que significaría que hay ciertos sujetos de los que no se tiene información, los cuales convendría eliminar del análisis.

In [21]:
#con axis=1 se analizan las filas y con all comprobamos que todos los nulos estén en la misma fila
#no tenemos en cuenta la columna de person_id porque entonces nos va a devolver false para todas las filas, puesto que la columna de id no tiene valores nulos
diet_df.drop(columns=["person_id"]).isnull().all(axis=1).value_counts() 

False    8531
True     1282
Name: count, dtype: int64

El resultado confirma que hay 8531 filas en las que no todos los valores son nulos, y 1282 filas, que representan 1282 sujetos, en las que todos los valores son nulos. Se eliminan estas filas para garantizar un análisis con datos reales. 

In [22]:
#eliminamos los nulos y comprobamos
diet_df = diet_df.dropna()
diet_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8531 entries, 0 to 9812
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   person_id                    8531 non-null   int64  
 1   kcal                         8531 non-null   float64
 2   protein_g                    8531 non-null   float64
 3   fat_g                        8531 non-null   float64
 4   cholesterol_mg               8531 non-null   float64
 5   carb_g                       8531 non-null   float64
 6   sugar_g                      8531 non-null   float64
 7   fiber_g                      8531 non-null   float64
 8   alcohol_g                    8531 non-null   float64
 9   total_monosat_fatty_acids_g  8531 non-null   float64
 10  total_polysat_fatty_acids_g  8531 non-null   float64
 11  calcium_mg                   8531 non-null   float64
 12  magnesium_mg                 8531 non-null   float64
dtypes: float64(12), int64(1

### 2.3 Comprobación de valores únicos

In [23]:
diet_df['person_id'].nunique()

8531

## 3. Limpieza del dataset examination

In [24]:
# Ánalisis general: visualización de las primeras filas para la comprensión de datos
exam_df.head()

,SEQN,PEASCST1,PEASCTM1,PEASCCT1,BPXCHR,BPAARM,BPACSZ,BPXPLS,BPXPULS,BPXPTY,...,CSXLEAOD,CSXSOAOD,CSXGRAOD,CSXONOD,CSXNGSOD,CSXSLTRT,CSXSLTRG,CSXNART,CSXNARG,CSAEFFRT
0,73557,1,620.0,NaN,NaN,1.0,4.0,86.0,1.0,1.0,...,2.0,1.0,1.0,1.0,4.0,62.0,1.0,NaN,NaN,1.0
1,73558,1,766.0,NaN,NaN,1.0,4.0,74.0,1.0,1.0,...,3.0,1.0,2.0,3.0,4.0,28.0,1.0,NaN,NaN,1.0
2,73559,1,665.0,NaN,NaN,1.0,4.0,68.0,1.0,1.0,...,2.0,1.0,2.0,3.0,4.0,49.0,1.0,NaN,NaN,3.0
3,73560,1,803.0,NaN,NaN,1.0,2.0,64.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,73561,1,949.0,NaN,NaN,1.0,3.0,92.0,1.0,1.0,...,3.0,1.0,4.0,3.0,4.0,NaN,NaN,NaN,NaN,1.0


### 3.1 Transformación de columnas con datos médicos

In [25]:
#obtenemos las columnas
exam_df.columns

Index(['SEQN', 'PEASCST1', 'PEASCTM1', 'PEASCCT1', 'BPXCHR', 'BPAARM',
       'BPACSZ', 'BPXPLS', 'BPXPULS', 'BPXPTY',
       ...
       'CSXLEAOD', 'CSXSOAOD', 'CSXGRAOD', 'CSXONOD', 'CSXNGSOD', 'CSXSLTRT',
       'CSXSLTRG', 'CSXNART', 'CSXNARG', 'CSAEFFRT'],
      dtype='object', length=224)

Al igual que en la limpieza de los dataset anteriores, se renombran y seleccionan aquellas columnas relevantes para el análisis.

In [26]:
#se renombran las columnas
exam_df=exam_df.rename(columns={
    "SEQN": "person_id",
    'BMXWT': 'weight_kg',
    'BMXHT': 'height_cm',
    'BMXBMI': 'body_mass_index',
    'BMXWAIST': 'waist_cm',
    'BPXSY1': 'systolic_blood_pressure',
    'BPXDI1': 'diastolic_bp',
    'BPXPLS': 'pulse'
      })

In [27]:
exam_df=exam_df[[
    "person_id",
    'weight_kg',
    'height_cm',
    'body_mass_index',
    'waist_cm', 
    'systolic_blood_pressure',
    'diastolic_bp',
    'pulse'
]]

In [28]:
#comprobamos
exam_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   person_id                9813 non-null   int64  
 1   weight_kg                9723 non-null   float64
 2   height_cm                9067 non-null   float64
 3   body_mass_index          9055 non-null   float64
 4   waist_cm                 8661 non-null   float64
 5   systolic_blood_pressure  7172 non-null   float64
 6   diastolic_bp             7172 non-null   float64
 7   pulse                    7549 non-null   float64
dtypes: float64(7), int64(1)
memory usage: 613.4 KB


### 3.2 Tratamiento de valores nulos

El output anterior muestra la existencia de valores nulos repartidos en todas las olumnas. Como todas las columnas contienen datos numéricos, se realiza una imputación con la mediana.

In [29]:
#imputamos con la mediana en las columnas numéricas
#para no ir una por una creamos una lista
exam_column=[  'weight_kg',
    'height_cm',
    'body_mass_index',
    'waist_cm', 
    'systolic_blood_pressure',
    'diastolic_bp',
    'pulse']
exam_df[exam_column] = exam_df[exam_column].fillna(exam_df[exam_column].median())

In [30]:
#comprobamos
exam_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9813 entries, 0 to 9812
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   person_id                9813 non-null   int64  
 1   weight_kg                9813 non-null   float64
 2   height_cm                9813 non-null   float64
 3   body_mass_index          9813 non-null   float64
 4   waist_cm                 9813 non-null   float64
 5   systolic_blood_pressure  9813 non-null   float64
 6   diastolic_bp             9813 non-null   float64
 7   pulse                    9813 non-null   float64
dtypes: float64(7), int64(1)
memory usage: 613.4 KB


### 3.3 Comprobación de los valores únicos

In [31]:
exam_df['person_id'].nunique()

9813

## 4. Limpieza del dataset question

In [32]:
#visualización general de los datos
question_df.head()

,SEQN,ACD011A,ACD011B,ACD011C,ACD040,ACD110,ALQ101,ALQ110,ALQ120Q,ALQ120U,...,WHD080U,WHD080L,WHD110,WHD120,WHD130,WHD140,WHQ150,WHQ030M,WHQ500,WHQ520
0,73557,1.0,NaN,NaN,NaN,NaN,1.0,NaN,1.0,3.0,...,NaN,40.0,270.0,200.0,69.0,270.0,62.0,NaN,NaN,NaN
1,73558,1.0,NaN,NaN,NaN,NaN,1.0,NaN,7.0,1.0,...,NaN,NaN,240.0,250.0,72.0,250.0,25.0,NaN,NaN,NaN
2,73559,1.0,NaN,NaN,NaN,NaN,1.0,NaN,0.0,NaN,...,NaN,NaN,180.0,190.0,70.0,228.0,35.0,NaN,NaN,NaN
3,73560,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.0,3.0,3.0
4,73561,1.0,NaN,NaN,NaN,NaN,1.0,NaN,0.0,NaN,...,NaN,NaN,150.0,135.0,67.0,170.0,60.0,NaN,NaN,NaN


### 4.1 Transformación de las columnas de cuestionario

In [33]:
#obtenemos las columnas
question_df.columns

Index(['SEQN', 'ACD011A', 'ACD011B', 'ACD011C', 'ACD040', 'ACD110', 'ALQ101',
       'ALQ110', 'ALQ120Q', 'ALQ120U',
       ...
       'WHD080U', 'WHD080L', 'WHD110', 'WHD120', 'WHD130', 'WHD140', 'WHQ150',
       'WHQ030M', 'WHQ500', 'WHQ520'],
      dtype='object', length=953)

Se seleccionan y renombran las columnas más relevantes.

In [34]:
#se renombran las columnas
question_df=question_df.rename(columns={
    "SEQN": "person_id",
    "PAQ620": "moderate_activity_day",
    "PAQ605": "vigorous_activity_day",
    "ALQ120U": "alcohol_frequency",
    "SLD010H": "sleep_hours",
    "WHQ030M": "self_reported_weight",
    "WHQ500": "trying_to_do_abt_weight"
      })

In [35]:
question_df=question_df[[
    "person_id",
    "moderate_activity_day",
    "vigorous_activity_day",
    "alcohol_frequency",
    "sleep_hours",
    "self_reported_weight",
    "trying_to_do_abt_weight"
]]

In [36]:
question_df.head()

,person_id,moderate_activity_day,vigorous_activity_day,alcohol_frequency,sleep_hours,self_reported_weight,trying_to_do_abt_weight
0,73557,2.0,2.0,3.0,7.0,NaN,NaN
1,73558,1.0,2.0,1.0,9.0,NaN,NaN
2,73559,1.0,2.0,NaN,8.0,NaN,NaN
3,73560,NaN,NaN,NaN,NaN,3.0,3.0
4,73561,1.0,2.0,NaN,9.0,NaN,NaN


Al igual que con el primer datset se realiza un mapeo de los datos para interpretar adecuadamente los valores.

In [37]:
# Mapeamos códigos a texto
question_df['moderate_activity_day']=question_df['moderate_activity_day'].map({
    1: 'Yes',
    2: 'No',
    7: 'Refused',
    9: 'Dont know'
})
question_df['vigorous_activity_day']=question_df['vigorous_activity_day'].map({
    1: 'Yes',
    2: 'No',
    7: 'Refused',
    9: 'Dont know'
})
question_df["self_reported_weight"]=question_df["self_reported_weight"].map({
    1: 'Fat or overweight',
    2: 'Too thin',
    3: 'About the right weight',
    7: 'Refused',
    9: 'Dont know'
})
question_df["trying_to_do_abt_weight"]=question_df["trying_to_do_abt_weight"].map({
    1: 'Lose weight',
    2: 'Gain weight',
    3: 'Stay the same weight',
    4: 'Not trying to do anything about your weight',
    7: 'Refused',
    9: 'Dont know'
})
question_df["alcohol_frequency"]=question_df["alcohol_frequency"].map({
    1: 'Weekly',
    2: 'Monthly',
    3: 'Yearly',
    7: 'Refused',
    9: 'Dont know'
})
question_df["sleep_hours"] = question_df["sleep_hours"].replace(99, np.nan)


In [38]:
question_df.head()

,person_id,moderate_activity_day,vigorous_activity_day,alcohol_frequency,sleep_hours,self_reported_weight,trying_to_do_abt_weight
0,73557,No,No,Yearly,7.0,NaN,NaN
1,73558,Yes,No,Weekly,9.0,NaN,NaN
2,73559,Yes,No,NaN,8.0,NaN,NaN
3,73560,NaN,NaN,NaN,NaN,About the right weight,Stay the same weight
4,73561,Yes,No,NaN,9.0,NaN,NaN


In [39]:
question_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10175 entries, 0 to 10174
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   person_id                10175 non-null  int64  
 1   moderate_activity_day    7148 non-null   object 
 2   vigorous_activity_day    7148 non-null   object 
 3   alcohol_frequency        3593 non-null   object 
 4   sleep_hours              6454 non-null   float64
 5   self_reported_weight     1478 non-null   object 
 6   trying_to_do_abt_weight  1478 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 556.6+ KB


### 4.2 Tratamiento de valores nulos

Para el tratamiento de valores nulos, se realiza una imputación con 'unknown' en las columnas categóricas y con la mediana para las columnas numéricas. 

In [40]:
#imputamos con unknown en las columnas categóricas
quest_categoric_column=["moderate_activity_day",
    "vigorous_activity_day",
    "alcohol_frequency",
    "self_reported_weight",
    "trying_to_do_abt_weight"]
question_df[quest_categoric_column] = question_df[quest_categoric_column].fillna('Unknown')

#imputamos con la mediana en las columnas numéricas
question_df["sleep_hours"] = question_df['sleep_hours'].fillna(question_df['sleep_hours'].median())

In [41]:
# y comprobamos
question_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10175 entries, 0 to 10174
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   person_id                10175 non-null  int64  
 1   moderate_activity_day    10175 non-null  object 
 2   vigorous_activity_day    10175 non-null  object 
 3   alcohol_frequency        10175 non-null  object 
 4   sleep_hours              10175 non-null  float64
 5   self_reported_weight     10175 non-null  object 
 6   trying_to_do_abt_weight  10175 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 556.6+ KB


### 4.3 Comprobación de valores únicos

In [42]:
question_df['person_id'].nunique()

10175

## 5. Unión de los datasets

Una vez terminada la fase de carga y limpieza, se realiza la unión de los cuatro datasets. Para ello, se emplean uniones de tipo inner join lo que garantiza que el dataset final incluya únicamente participantes con información completa en todos los conjuntos de datos, facilitando un análisis consistente.

In [43]:
nutrition_df= demog_df.merge(diet_df, on='person_id', how='inner')
nutrition_df= nutrition_df.merge(exam_df, on='person_id', how='inner')
nutrition_df= nutrition_df.merge(question_df, on='person_id', how='inner')


In [44]:
# revisión del dataframe final 
nutrition_df.info()  #verificamos los nulos y tipos de datos en cada columna
nutrition_df.head()  #visualizamos las 5 primeras filas


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8531 entries, 0 to 8530
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   person_id                    8531 non-null   int64  
 1   gender                       8531 non-null   object 
 2   age                          8531 non-null   int64  
 3   race                         8531 non-null   object 
 4   country                      8531 non-null   object 
 5   education_level              8531 non-null   object 
 6   marital_status               8531 non-null   object 
 7   pregnancy_status             8531 non-null   object 
 8   household_size               8531 non-null   int64  
 9   household_income             8531 non-null   object 
 10  family_income                8531 non-null   float64
 11  family_poverty_ratio         8531 non-null   float64
 12  kcal                         8531 non-null   float64
 13  protein_g         

,person_id,gender,age,race,country,education_level,marital_status,pregnancy_status,household_size,household_income,...,waist_cm,systolic_blood_pressure,diastolic_bp,pulse,moderate_activity_day,vigorous_activity_day,alcohol_frequency,sleep_hours,self_reported_weight,trying_to_do_abt_weight
0,73557,Male,69,Non-Hispanic Black,USA,High school/GED,Separated,Unknown,3,"15,000 to 19,999 $",...,100.0,122.0,72.0,86.0,No,No,Yearly,7.0,Unknown,Unknown
1,73558,Male,54,Non-Hispanic White,USA,High school/GED,Married,Unknown,4,"35,000 to 44,999 $",...,107.6,156.0,62.0,74.0,Yes,No,Weekly,9.0,Unknown,Unknown
2,73559,Male,72,Non-Hispanic White,USA,Some college/AA,Married,Unknown,2,"65,000 to 74,999 $",...,109.2,140.0,90.0,68.0,Yes,No,Unknown,8.0,Unknown,Unknown
3,73560,Male,9,Non-Hispanic White,USA,Unknown,Unknown,Unknown,4,"55,000 to 64,999 $",...,61.0,108.0,38.0,64.0,Unknown,Unknown,Unknown,7.0,About the right weight,Stay the same weight
4,73561,Female,73,Non-Hispanic White,USA,College graduate or above,Married,Unknown,2,"100,000 $ and Over",...,87.8,136.0,86.0,92.0,Yes,No,Unknown,9.0,Unknown,Unknown


In [45]:
nutrition_df.to_csv("C:\\Users\\al_ce\\Documents\\CELIA\\master data analytics\\PROYECTO_FINAL\\DATA\\nutririon_processed\\nutrition_clean.csv", index=False)
